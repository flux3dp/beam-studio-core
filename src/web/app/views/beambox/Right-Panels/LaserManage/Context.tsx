import React, { Dispatch } from 'react';

import i18n from 'helpers/i18n';
import isObjectEmpty from 'helpers/is-object-empty';
import storage from 'implementations/storage';
import { getDefaultPresetData } from 'helpers/presets/preset-helper';
import { ILaserConfig, ILaserDataChanges } from 'interfaces/ILaserConfig';

interface State {
  configs: ILaserConfig[];
  selectedItem : {
    name: string;
    isCustomized: boolean;
  };
  presetsInUse: { [name: string]: boolean };
  dataChanges: { [key: string]: ILaserDataChanges };
  displayValues: {
    speed: number;
    power: number;
    repeat: number;
    zStep: number;
  };
}

interface SelectPayload {
  name: string;
  isCustomized: boolean;
}
interface ChangePayload {
  name: string;
  key: 'speed' | 'power' | 'repeat' | 'zStep';
  value: number;
}
interface AddConfigPayload {
  name: string;
}
interface SwapConfigPayload {
  orig: number;
  dist: number;
}

type Action = {
  type: 'select';
  payload: SelectPayload;
} | {
  type: 'change';
  payload: ChangePayload;
} | {
  type: 'add-config';
  payload: AddConfigPayload;
} | {
  type: 'swap-config';
  payload: SwapConfigPayload;
} | {
  type: 'add-preset';
  payload?: never;
} | {
  type: 'remove-preset';
  payload?: never;
} | {
  type: 'delete';
  payload?: never;
} | {
  type: 'reset';
  payload?: never;
};

export const getInitState = (selectingConfigName: string): State => {
  const configs: ILaserConfig[] = storage.get('customizedLaserConfigs') || [];
  const presetsInUse = storage.get('defaultLaserConfigsInUse');
  const selectedConfig = configs.find((c) => c.name === selectingConfigName);
  const { speed, power, repeat, zStep } = selectedConfig || {};
  return {
    configs,
    selectedItem: { name: selectingConfigName, isCustomized: true },
    presetsInUse,
    dataChanges: {},
    displayValues: {
      speed: speed || 1,
      power: power || 0,
      repeat: repeat || 1,
      zStep: zStep || 0,
    },
  };
};

const selectAction = (state: State, payload: SelectPayload): State => {
  const { name, isCustomized } = payload;
  if (isCustomized) {
    const { configs, dataChanges } = state;
    const selectedConfig = configs.find((c) => c.name === name);
    const { speed, power, repeat, zStep } = { ...selectedConfig, ...dataChanges[name] };
    return {
      ...state,
      selectedItem: { name, isCustomized },
      displayValues: { speed, power, repeat, zStep },
    };
  }
  // selecting presets
  const { speed, power, repeat } = getDefaultPresetData(name);
  const zStep = 0;
  return {
    ...state,
    selectedItem: { name, isCustomized },
    displayValues: { speed, power, repeat, zStep },
  };
};

const changeAction = (state: State, payload: ChangePayload): State => {
  const { name, key, value } = payload;
  const { configs, dataChanges, displayValues } = state;
  const selectedConfig = configs.find((c) => c.name === name);
  if (selectedConfig[key] !== value) {
    if (!dataChanges[name]) {
      const change = {};
      change[key] = value;
      dataChanges[name] = change;
    } else {
      dataChanges[name][key] = value;
    }
  } else {
    if (dataChanges[name]) {
      delete dataChanges[name][key];
    }
    if (isObjectEmpty(dataChanges[name])) {
      delete dataChanges[name];
    }
  }
  return {
    ...state,
    dataChanges,
    displayValues: {
      ...displayValues,
      [key]: value,
    },
  };
};

const addConfigAction = (state: State, payload: AddConfigPayload): State => {
  const { name } = payload;
  const { configs } = state;
  configs.push({ name, speed: 20, power: 15, repeat: 1, zStep: 0 });
  return {
    ...state,
    selectedItem: { name, isCustomized: true },
    displayValues: { speed: 20, power: 15, repeat: 1, zStep: 0 },
  };
};

const swapConfigAction = (state: State, payload: SwapConfigPayload): State => {
  const { orig, dist } = payload;
  const { configs } = state;
  [configs[orig], configs[dist]] = [configs[dist], configs[orig]];
  return {
    ...state,
    configs,
  };
};

const addPresetAction = (state: State): State => {
  const { selectedItem: { name: key, isCustomized }, configs, presetsInUse } = state;
  if (!isCustomized && key !== '') {
    const unit = storage.get('default-units') || 'mm';
    const displayName = i18n.lang.beambox.right_panel.laser_panel.dropdown[unit][key];
    if (presetsInUse[key]) {
      return { ...state, selectedItem: { name: displayName, isCustomized: true } };
    }
    const { speed, power, repeat } = getDefaultPresetData(key);
    presetsInUse[key] = true;
    configs.push({ name: displayName, speed, power, repeat, zStep: 0, isDefault: true, key });
    return {
      ...state,
      selectedItem: { name: displayName, isCustomized: true },
      configs,
      presetsInUse,
    };
  }
  return state;
};

const removePresetAction = (state: State): State => {
  const { selectedItem: { name, isCustomized }, configs, presetsInUse, dataChanges } = state;
  const unit = storage.get('default-units') || 'mm';
  const configName = isCustomized ? name : i18n.lang.beambox.right_panel.laser_panel.dropdown[unit][name];
  const index = configs.findIndex((c) => c.name === configName);

  if (index < 0 || !configs[index].isDefault) return state;
  const { key } = configs[index];
  presetsInUse[key] = false;
  configs.splice(index, 1);
  if (configs.length > 0) {
    // if custom config list is not empty, selecting config above
    const newIndex = Math.min(index, configs.length - 1);
    const newConfig = configs[newIndex];
    const { name: newConfgName, speed, power, repeat, zStep } = { ...newConfig, ...dataChanges[newConfig.name] };
    return {
      ...state,
      configs,
      presetsInUse,
      selectedItem: { name: newConfgName, isCustomized: true },
      displayValues: { speed, power, repeat: repeat || 1, zStep: zStep || 0 },
    };
  }
  // selecting remove preset in preset list
  return {
    ...state,
    configs,
    presetsInUse,
    selectedItem: { name, isCustomized: false },
  };
};

const deleteAction = (state: State): State => {
  const { selectedItem: { name, isCustomized }, configs, dataChanges } = state;
  if (!isCustomized) return state;
  const idx = configs.findIndex((c) => c.name === name);
  if (idx < 0) return state;
  if (configs[idx].isDefault) return removePresetAction(state);
  configs.splice(idx, 1);
  if (configs.length > 0) {
    // if custom config list is not empty, selecting config above
    const newIndex = Math.min(idx, configs.length - 1);
    const newConfig = configs[newIndex];
    const { name: newConfgName, speed, power, repeat, zStep } = { ...newConfig, ...dataChanges[newConfig.name] };
    return {
      ...state,
      configs,
      selectedItem: { name: newConfgName, isCustomized: true },
      displayValues: { speed, power, repeat: repeat || 1, zStep: zStep || 0 },
    };
  }
  return {
    ...state,
    configs,
    selectedItem: { name: '', isCustomized: false },
    displayValues: { speed: 1, power: 0, repeat: 1, zStep: 0 },
  };
};

const resetAction = (state: State): State => {
  const configs = storage.get('customizedLaserConfigs') || [];
  const presetsInUse = storage.get('defaultLaserConfigsInUse') || {};
  return { ...state, configs, presetsInUse };
};

export const reducer = (state: State, action: Action): State => {
  const { type, payload } = action;
  switch (type) {
    case 'select':
      return selectAction(state, payload);
    case 'change':
      return changeAction(state, payload);
    case 'add-config':
      return addConfigAction(state, payload);
    case 'swap-config':
      return swapConfigAction(state, payload);
    case 'add-preset':
      return addPresetAction(state);
    case 'remove-preset':
      return removePresetAction(state);
    case 'delete':
      return deleteAction(state);
    case 'reset':
      return resetAction(state);
    default:
      break;
  }
  return state;
};

interface Context {
  state: State;
  dispatch: Dispatch<Action>;
}

export default React.createContext<Context>({} as Context);
