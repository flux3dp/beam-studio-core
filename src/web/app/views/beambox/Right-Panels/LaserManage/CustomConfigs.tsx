import classNames from 'classnames';
import React, { memo, useContext, useEffect, useRef } from 'react';
import { Button, Col } from 'antd';
import { PlusCircleFilled } from '@ant-design/icons';

import alertCaller from 'app/actions/alert-caller';
import dialogCaller from 'app/actions/dialog-caller';
import useI18n from 'helpers/useI18n';
import { ILaserConfig } from 'interfaces/ILaserConfig';

import Context from './Context';
import styles from './ConfigList.module.scss';

const CustomConfigs = (): JSX.Element => {
  const { state, dispatch } = useContext(Context);
  const listRef = useRef<HTMLDivElement>(null);
  const draggingIndex = useRef(-1);
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;
  const {
    selectedItem: { name, isCustomized },
    configs,
    dataChanges,
  } = state;

  useEffect(() => {
    if (isCustomized) {
      const idx = configs.findIndex((c) => c.name === name);
      if (idx >= 0 && listRef.current) {
        const list = listRef.current;
        const item = listRef.current.childNodes.item(idx) as HTMLDivElement;
        const itemPos = item.offsetTop - list.offsetTop - list.clientTop;
        if (itemPos + item.scrollHeight < list.scrollTop || itemPos > list.scrollTop + list.offsetHeight) {
          list.scrollTop = itemPos;
        }
      }
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [name, isCustomized]);

  const handleAddClick = () => {
    dialogCaller.promptDialog({
      caption: t.new_config_name,
      defaultValue: '',
      onYes: (newName) => {
        if (!newName) return;
        const isPresetNameUsed = configs.some((c) => c.name === newName);
        if (isPresetNameUsed) alertCaller.popUpError({ message: t.existing_name });
        else dispatch({ type: 'add-config', payload: { name: newName } });
      },
    });
  };

  const handleDragStart = (config: ILaserConfig, idx: number) => {
    draggingIndex.current = idx;
    dispatch({ type: 'select', payload: { name: config.name, isCustomized: true } });
  };

  const handleDragOver = (idx: number) => {
    if (draggingIndex.current >= 0 && draggingIndex.current !== idx) {
      dispatch({ type: 'swap-config', payload: { orig: draggingIndex.current, dist: idx } });
      draggingIndex.current = idx;
    }
  };

  const handleDragEnd = () => {
    draggingIndex.current = -1;
  };

  return (
    <Col span={11}>
      <div className={styles.title}>
        <strong>
          {t.customized}
          <Button type="text" onClick={handleAddClick}>
            <PlusCircleFilled />
          </Button>
        </strong>
      </div>
      <div id="custom-config-list" className={styles.list} ref={listRef} onDragOver={(e) => e.preventDefault()}>
        {configs.map((config, i) => {
          const hasChange = !!dataChanges[config.name];
          return (
            <div
              draggable
              key={config.name}
              className={classNames(styles.item, {
                [styles.selected]: isCustomized && name === config.name,
                [styles.noborder]: configs.length >= 8 && i === configs.length - 1,
              })}
              onClick={() => dispatch({ type: 'select', payload: { name: config.name, isCustomized: true } })}
              onDragStart={() => handleDragStart(config, i)}
              onDragOver={() => handleDragOver(i)}
              onDragEnd={handleDragEnd}
            >
              <div className={styles.name}>{`${config.name + (hasChange ? ' *' : '')}`}</div>
              {config.isDefault ? <span className={styles.sub}>{t.default}</span> : null}
            </div>
          );
        })}
      </div>
    </Col>
  );
};

export default memo(CustomConfigs);
