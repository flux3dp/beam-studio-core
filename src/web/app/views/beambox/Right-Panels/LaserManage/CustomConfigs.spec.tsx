import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import Context from './Context';
import CustomConfigs from './CustomConfigs';

jest.mock('app/views/beambox/Right-Panels/LaserManage/Context', () => React.createContext({}));

const mockPopUpError = jest.fn();
jest.mock('app/actions/alert-caller', () => ({
  popUpError: (...args) => mockPopUpError(...args),
}));

const mockPromptDialog = jest.fn();
jest.mock('app/actions/dialog-caller', () => ({
  promptDialog: (...args) => mockPromptDialog(...args),
}));

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    right_panel: {
      laser_panel: {
        customized: 'customized',
        default: 'default',
      },
    },
  },
}));

const mockDispatch = jest.fn();
describe('test PresetsList', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    const mockState = {
      selectedItem: { name: 'preset1', isCustomized: true },
      configs: [{ name: 'preset1', isDefault: true }, { name: 'preset2', isDefault: false }],
      dataChanges: { preset2: { speed: 10 } },
    };
    const { container } = render(
      <Context.Provider value={{ state: mockState as any, dispatch: mockDispatch }}>
        <CustomConfigs />
      </Context.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('selecting should work', () => {
    const mockState = {
      selectedItem: { name: 'preset1', isCustomized: true },
      configs: [{ name: 'preset1', isDefault: true }, { name: 'preset2', isDefault: false }],
      dataChanges: { preset2: { speed: 10 } },
    };
    const { getByText } = render(
      <Context.Provider value={{ state: mockState as any, dispatch: mockDispatch }}>
        <CustomConfigs />
      </Context.Provider>
    );
    expect(mockDispatch).not.toBeCalled();
    fireEvent.click(getByText('preset2 *'));
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({
      type: 'select', payload: { name: 'preset2', isCustomized: true }
    });
  });

  test('add preset button should work', () => {
    const mockState = {
      selectedItem: { name: 'preset1', isCustomized: true },
      configs: [{ name: 'preset1', isDefault: true }, { name: 'preset2', isDefault: false }],
      dataChanges: { preset2: { speed: 10 } },
    };
    const { container } = render(
      <Context.Provider value={{ state: mockState as any, dispatch: mockDispatch }}>
        <CustomConfigs />
      </Context.Provider>
    );
    const addBtn = container.querySelector('button');
    expect(mockPromptDialog).not.toBeCalled();
    fireEvent.click(addBtn);
    expect(mockPromptDialog).toBeCalledTimes(1);
    const { onYes } = mockPromptDialog.mock.calls[0][0];
    expect(mockPopUpError).not.toBeCalled();
    onYes('preset2');
    expect(mockPopUpError).toBeCalledTimes(1);
    expect(mockDispatch).not.toBeCalled();
    onYes('preset3');
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith({ type: 'add-config', payload: { name: 'preset3' } });
  });
});
