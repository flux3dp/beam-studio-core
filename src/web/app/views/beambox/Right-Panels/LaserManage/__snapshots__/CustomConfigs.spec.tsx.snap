// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PresetsList should render correctly 1`] = `
<div>
  <div
    class="ant-col ant-col-11 css-dev-only-do-not-override-hash"
  >
    <div
      class="title"
    >
      <strong>
        customized
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-text"
          type="button"
        >
          <span
            aria-label="plus-circle"
            class="anticon anticon-plus-circle"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm192 472c0 4.4-3.6 8-8 8H544v152c0 4.4-3.6 8-8 8h-48c-4.4 0-8-3.6-8-8V544H328c-4.4 0-8-3.6-8-8v-48c0-4.4 3.6-8 8-8h152V328c0-4.4 3.6-8 8-8h48c4.4 0 8 3.6 8 8v152h152c4.4 0 8 3.6 8 8v48z"
              />
            </svg>
          </span>
          
        </button>
      </strong>
    </div>
    <div
      class="list"
      id="custom-config-list"
    >
      <div
        class="item selected"
        draggable="true"
      >
        <div
          class="name"
        >
          preset1
        </div>
        <span
          class="sub"
        >
          default
        </span>
      </div>
      <div
        class="item"
        draggable="true"
      >
        <div
          class="name"
        >
          preset2 *
        </div>
      </div>
    </div>
  </div>
</div>
`;
